src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,522 @@
+                }
+                throw new RuntimeException(ERROR_MESSAGE);
+            }
+        }
+        catch (InterruptedException e) {
+            log.error(ERROR_MESSAGE);
+            executorService.shutdownNow();
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+
+        evaluateErrorMap(errorMap);
+    }
+
+    /**
+     * Executes the migration for a batch of programming exercises. Expected to be in a separate thread.
+     *
+     * @param start         The start index of the batch page
+     * @param end           The end index of the batch page
+     * @param exerciseCount The total number of exercises for this thread
+     * @param errorMap      The error map to write errors to
+     */
+    private void executeBatchedMigration(long start, long end, long exerciseCount, Map<ProgrammingExercise, Boolean> errorMap) {
+        // do batches
+        for (int batchStartIndex = (int) start, batch = 0; batchStartIndex < end; batchStartIndex++, batch = batch + BATCH_SIZE) {
+            var batchStart = System.currentTimeMillis();
+            Pageable pageable = PageRequest.of(batchStartIndex, BATCH_SIZE);
+
+            var templateParticipationPage = templateProgrammingExerciseParticipationRepository.findAll(pageable);
+            log.info("Found {} template programming exercises to migrate in batch", templateParticipationPage.getTotalElements());
+            templateParticipationPage.map(templateParticipation -> {
+                boolean buildPlanExists = ciMigrationService.orElseThrow().buildPlanExists(templateParticipation.getProgrammingExercise().getProjectKey(),
+                        templateParticipation.getProgrammingExercise().getTemplateBuildPlanId());
+                if (!buildPlanExists) {
+                    log.warn("Build plan {} does not exist for exercise {}", templateParticipation.getProgrammingExercise().getTemplateBuildPlanId(),
+                            templateParticipation.getProgrammingExercise().getId());
+                    log.info("Skipping migration for template of exercise {} and setting build_plan_id to null", templateParticipation.getProgrammingExercise().getId());
+                    templateParticipation.setBuildPlanId(null);
+                    templateProgrammingExerciseParticipationRepository.save(templateParticipation);
+                    return null;
+                }
+                log.info("Migrating build plan with name {} for exercise {}",
+                        templateParticipation.getProgrammingExercise().getProjectKey() + "-" + templateParticipation.getBuildPlanId(),
+                        templateParticipation.getProgrammingExercise().getId());
+                return templateParticipation;
+            }).filter(Objects::nonNull).map(templateParticipation -> {
+                String branch = templateParticipation.getProgrammingExercise().getBranch();
```

Comment:
You should only use the getter if you are sure that the value is already stored in Artemis. This is also explained in the java doc of the getter. You should use `getOrRetrieveBranchOfExercise` in the AbstractVersionControllService instead